pkg/client/src/app/pages/applications/analysis-wizard/components/add-custom-rules.tsx

```diff
@@ -71,37 +77,73 @@ export const AddCustomRules: React.FunctionComponent<IAddCustomRules> = ({
     setReadFileData(newReadFiles);
   };
 
-  // callback that will be called by the react dropzone with the newly dropped file objects
+  const isSchemaValid = (value: string) => {
+    const validationResult = xmllint.xmllint.validateXML({
+      xml: value,
+      schema: XSDSchema,
+    });
+
+    if (!validationResult.errors) return true;
+    return false;
+  };
+
+  const validateXMLFile = (data: string) => {
+    // Filter out "data:text/xml;base64," from data
+    const payload = atob(data.substring(21));
+
+    let isXML = false;
+    let isXSD = false;
+
+    const validationObject = XMLValidator.validate(payload, {
+      allowBooleanAttributes: true,
+    });
+
+    if (validationObject === true) {
+      isXML = true;
+      if (isSchemaValid(payload)) {
+        isXSD = true;
+      }
+    }
+    console.log("XML :", isXML);
+    console.log("XSD :", isXSD);
+  };
+
   const handleFileDrop = (droppedFiles: File[]) => {
-    // identify what, if any, files are re-uploads of already uploaded files
     const currentFileNames = currentFiles.map((file) => file.name);
     const reUploads = droppedFiles.filter((droppedFile) =>
       currentFileNames.includes(droppedFile.name)
     );
 
-    /** this promise chain is needed because if the file removal is done at the same time as the file adding react
-     * won't realize that the status items for the re-uploaded files needs to be re-rendered */
     Promise.resolve()
       .then(() => removeFiles(reUploads.map((file) => file.name)))
       .then(() =>
         setCurrentFiles((prevFiles: File[]) => [...prevFiles, ...droppedFiles])
       );
   };
 
-  // callback called by the status item when a file is successfully read with the built-in file reader
   const handleReadSuccess = (data: string, file: File) => {
-    setReadFileData((prevReadFiles) => [
-      ...prevReadFiles,
-      { data, fileName: file.name, loadResult: "success" },
-    ]);
+    validateXMLFile(data);
+
+    const fileList = [
+      ...readFileData,
+      { data, fileName: file.name, loadResult: "success" } as IReadFile,
```

Comment:
Why are we asserting here?

Reply:
Because the object is created on the fly. 
I suppose I could create a sub-type or assert

Reply:
https://medium.com/swlh/why-you-should-avoid-type-assertions-in-typescript-5494e3d04dd

Reply:
That makes sense and this is really important and trying to write lighter code defeated the purpose of type in this case. Thank you for bringing my attention to this.